src/ophyd_async/core/detector.py

```diff
@@ -179,25 +231,95 @@ async def trigger(self) -> None:
+        Prepare the detector with trigger information. This is determined at and passed
+        in from the plan level.
 
-        async for doc in self.writer.collect_stream_docs(indices_written):
-            yield doc
-        # async for doc in self.writer.collect_stream_docs(indices_written):
-        #     yield doc
+        This currently only prepares detectors for flyscans and stepscans just use the
+        trigger information determined in trigger.
+
+        To do: Unify prepare to be use for both fly and step scans.
+        """
+        assert type(value) is TriggerInfo
+        self._trigger_info = value
+        self._current_frame = current_frame
+
+        self._offset -= self._current_frame
+        self._current_frame = 0
+        self._last_frame = self._current_frame + self._trigger_info.num
+
+        await self.ensure_armed(self._trigger_info)
+
+    async def ensure_armed(self, trigger_info: TriggerInfo):
+        if (
+            not self._arm_status
+            or self._arm_status.done
+            or trigger_info != self._trigger_info
+        ):
+            # we need to re-arm
+            await self.controller.disarm()
+            required = self.controller.get_deadtime(trigger_info.livetime)
+            assert required <= trigger_info.deadtime, (
+                f"Detector {self.controller} needs at least {required}s deadtime, "
+                f"but trigger logic provides only {trigger_info.deadtime}s"
+            )
+            self._arm_status = await self.controller.arm(
+                num=trigger_info.num,
+                trigger=trigger_info.trigger,
+                exposure=trigger_info.livetime,
+            )
 
     @AsyncStatus.wrap
-    async def unstage(self) -> None:
-        """Stop data writing."""
-        await self.writer.close()
+    async def kickoff(self) -> None:
+        self._fly_status = AsyncStatus(self._fly(), self._watchers)
+        self._fly_start = time.monotonic()
+
+    async def _fly(self) -> None:
+        end_observation = self._last_frame - self._offset
+        await self._observe_writer_indicies(end_observation)
+
+    @AsyncStatus.wrap
+    async def complete(self) -> AsyncStatus:
+        assert self._fly_status, "Kickoff not run"
+        return await self._fly_status
+
+    async def describe_collect(self) -> Dict[str, Descriptor]:
+        return self._describe
+
+    async def collect_asset_docs(
+        self, index: Optional[int] = None
+    ) -> AsyncIterator[StreamAsset]:
+        """Collect stream datum documents for all indices written.
+
+        The index is optional, and provided for flyscans, however this needs to be
+        retrieved for stepscans.
+        """
+        if not index:
+            index = await self.writer.get_indices_written()
+        async for doc in self.writer.collect_stream_docs(index):
+            yield doc
+
+    async def get_index(self) -> int:
+        return await self.writer.get_indices_written()
```

Comment:
It felt a bit confusing at the time but if I remember correctly this is used for the protocol `WritesStreamAssets` to differentiate it from `WritesExternalAssets` as function arguments were not addressed in the runtime check, so the inclusion of an index in the `Collect_asset_docs` method is not sufficient and we need to make this distinction as we currently want to use both protocol in different areas. I agree this could probably be refactored, but wondering if that time is now.

Reply:
Of course! Can't believe I forgot that...